// http-message-signatures/lib/httpbis/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sign = exports.buildSignedData = exports.buildSignatureInputString = exports.extractComponent = exports.extractHeader = exports.defaultSigningComponents = void 0;
const url_1 = require("url");
exports.defaultSigningComponents = [
    '@method',
    '@path',
    '@query',
    '@authority',
    'content-type',
    'digest',
    'content-digest',
];
function extractHeader({ headers }, header, opts) {
    var _a, _b;
    const lcHeader = header.toLowerCase();
    const key = Object.keys(headers).find((name) => name.toLowerCase() === lcHeader);
    const allowMissing = (_a = opts === null || opts === void 0 ? void 0 : opts.allowMissing) !== null && _a !== void 0 ? _a : true;
    if (!allowMissing && !key) {
        throw new Error(`Unable to extract header "${header}" from message`);
    }
    let val = key ? (_b = headers[key]) !== null && _b !== void 0 ? _b : '' : '';
    if (Array.isArray(val)) {
        val = val.join(', ');
    }
    return val.toString().replace(/\s+/g, ' ');
}
exports.extractHeader = extractHeader;
function populateDefaultParameters(parameters) {
    return {
        created: new Date(),
        ...parameters,
    };
}
// see https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-message-signatures-06#section-2.3
function extractComponent(message, component) {
    switch (component) {
        case '@method':
            return message.method.toUpperCase();
        case '@target-uri':
            return message.url;
        case '@authority': {
            const url = new url_1.URL(message.url);
            const port = url.port ? parseInt(url.port, 10) : null;
            return `${url.host}${port && ![80, 443].includes(port) ? `:${port}` : ''}`;
        }
        case '@scheme': {
            const { protocol } = new url_1.URL(message.url);
            return protocol.slice(0, -1);
        }
        case '@request-target': {
            const { pathname, search } = new url_1.URL(message.url);
            return `${pathname}${search}`;
        }
        case '@path': {
            const { pathname } = new url_1.URL(message.url);
            return pathname;
        }
        case '@query': {
            const { search } = new url_1.URL(message.url);
            return search;
        }
        case '@status':
            if (!message.status) {
                throw new Error(`${component} is only valid for responses`);
            }
            return message.status.toString();
        case '@query-params':
        case '@request-response':
            throw new Error(`${component} is not implemented yet`);
        default:
            throw new Error(`Unknown specialty component ${component}`);
    }
}
exports.extractComponent = extractComponent;
function buildSignatureInputString(componentNames, parameters) {
    const components = componentNames.map((name) => `"${name.toLowerCase()}"`).join(' ');
    return `(${components})${Object.entries(parameters).map(([parameter, value]) => {
        if (typeof value === 'number') {
            return `;${parameter}=${value}`;
        }
        else if (value instanceof Date) {
            return `;${parameter}=${Math.floor(value.getTime() / 1000)}`;
        }
        else {
            return `;${parameter}="${value.toString()}"`;
        }
    }).join('')}`;
}
exports.buildSignatureInputString = buildSignatureInputString;
function buildSignedData(request, components, signatureInputString) {
    const parts = components.map((component) => {
        let value;
        if (component.startsWith('@')) {
            value = extractComponent(request, component);
        }
        else {
            value = extractHeader(request, component);
        }
        return `"${component.toLowerCase()}": ${value}`;
    });
    parts.push(`"@signature-params": ${signatureInputString}`);
    return parts.join('\n');
}
exports.buildSignedData = buildSignedData;
// @todo - should be possible to sign responses too
async function sign(request, opts) {
    var _a;
    const signingComponents = (_a = opts.components) !== null && _a !== void 0 ? _a : exports.defaultSigningComponents;
    const signingParams = populateDefaultParameters({
        ...opts.parameters,
        keyid: opts.keyId,
        alg: opts.signer.alg,
    });
    const signatureInputString = buildSignatureInputString(signingComponents, signingParams);
    const dataToSign = buildSignedData(request, signingComponents, signatureInputString);
    const signature = await opts.signer(Buffer.from(dataToSign));
    Object.assign(request.headers, {
        'Signature': `sig1=:${signature.toString('base64')}:`,
        'Signature-Input': `sig1=${signatureInputString}`,
    });
    return request;
}
exports.sign = sign;
